Add "Serialized transaction format" subsection to "Transaction Format"

IkerAlus opened the pull request:
- Added a subsection to "Transaction Format" section to add details about the actual format of a serialized transaction.
- Added the "Summary" subsection so it is clear the enumeration of the transaction process refers to the whole section.

The info for this update has been mainly take from [this example](https://github.com/paritytech/polkadot-interaction-examples-rs/blob/main/src/bin/05_transfer_balance.rs).

IkerAlus committed:
Add "Serialized transaction format" subsection to "Transaction Format"

- Added a subsection to "Transaction Format" section to add details about the actual format of a serialized transaction.
- Added the "Summary" subsection so it is clear the enumeration of the transaction process refers to the whole section.

The info for this update has been mainly take from [this example](https://github.com/paritytech/polkadot-interaction-examples-rs/blob/main/src/bin/05_transfer_balance.rs).

IkerAlus committed:
Beautify

joepetrowski commented on docs/build/build-transaction-construction.md:
```diff
@@ -35,6 +35,27 @@ function from the Balances pallet will take:
+- 7 bits: the transaction protocol version.
+- If there is a signature:
+  - a SCALE encoded `sp_runtime::MultiAddress::Id<AccountId32, u32>` indicating the signer(s) of the transaction.
+  - a SCALE encoded `sp_runtime::MultiSignature::S225519` with the signature.
```
Polkadot supports sr25519, ed25519, and ecdsa. This is just an example using sr25519 (with a typo :)), so the text should either indicate that or just make the example generic.
```suggestion
  - a SCALE encoded `sp_runtime::MultiSignature::{SigningScheme}` with the signature.
```

IkerAlus replied:
good point, updated and added a little clarification.

joepetrowski commented on docs/build/build-transaction-construction.md:
```diff
@@ -35,6 +35,27 @@ function from the Balances pallet will take:
+Before being submitted, transactions are serialized. Serialized transactions are hex encoded SCALE-encoded bytes with the following format:
+- Compact encoded number of SCALE-encoded bytes following this.
+- 1 bit: it is a 0 if no signature is present, or a 1 if it is.
+- 7 bits: the transaction protocol version.
```
This is different from the `transaction_version`. It's confusing, but at least note it here.

IkerAlus replied:
I though these 7 bits was exactly `transaction_version` as described [here](https://docs.substrate.io/v3/runtime/upgrades/).

What are they then? I will update this accordingly.

joepetrowski replied:
Nope, so this is a bit lower level, it's like the version of "what an extrinsic is" _to Substrate_, while `transaction_version` refers to backwards incompatible changes _to a specific runtime,_ namely when you break pallet/call indexing, function signatures (including types), or types in the serialized transaction itself (e.g. `AccountId` or `SignedExtension`s).

`transaction_version` came much later than `extrinsic_version`, and was a bit of a naming collision that most people just didn't notice. Unless there is a major major major change to Substrate, `extrinsic_version` will basically always `= 4`, and a change would mean basically needing to update everything.

You can see a few examples of its use in txwrapper:

- [Decoding extrinsics](https://github.com/paritytech/txwrapper-core/blob/d590b0912ee0bf9f357233024c064138cff36cd2/packages/txwrapper-core/src/core/decode/decodeSigningPayload.ts#L38) (notice `transaction_version` at L54)
- [The definition of a "method"](https://github.com/paritytech/txwrapper-core/blob/3cc7bf3e63f1e88d231454a98307afe4ab798b27/packages/txwrapper-core/src/core/method/defineMethod.ts#L106)
- [Explained a bit in an example](https://github.com/paritytech/txwrapper-core/blob/b3d7fc1ea49244cbda18d08f14a5d2cdd0b2e78c/packages/txwrapper-core/src/core/construct/createSigningPayload.ts#L27)

IkerAlus replied:
thanks a lot for the detailed context around this. 

I updated the text to "extrinsic version" and added the point that it should be 4.

IkerAlus replied:
I updated the description to "extrinsic version" also based on this PR: https://github.com/paritytech/substrate/commit/6bcc3851909284f934461e4dc45fec72ce1cb525

joepetrowski reviewed (commented):
Good start. Would add more contextual info, like that the actual serialization is defined in the runtime and can change. And at the end format for line length.

IkerAlus committed:
Address Joe's review

joepetrowski commented on docs/build/build-transaction-construction.md:
```diff
@@ -35,6 +35,29 @@ function from the Balances pallet will take:
+- If there is a signature:
+  - a SCALE encoded `sp_runtime::MultiAddress::Id<AccountId32, u32>` indicating the signer(s) of the transaction.
+  - a SCALE encoded `sp_runtime::MultiSignature::{SigningScheme}` with the signature\*.
+  - a SCALE encoded `sp_runtime::generic::Era` indicationg for how long this transaction will live in the pool.
```
Its being kicked from the pool is just a byproduct of it being invalid from the runtime's perspective (the pool won't include transactions that will _obviously_ fail, e.g. nonce too low, too old, etc.)
```suggestion
  - a SCALE encoded `sp_runtime::generic::Era` indicating for how long this transaction is valid.
```

salmad3 replied:
@IkerAlus

joepetrowski replied:
@DannyS03 - he's on vacation, he'll get back to it when he comes back.

IkerAlus replied:
good point, changed as suggested

IkerAlus committed:
Address Joe's points

joepetrowski commented on docs/build/build-transaction-construction.md:
```diff
@@ -35,6 +35,29 @@ function from the Balances pallet will take:
+- The specific transaction parameters or call data, which consists of:
+  - 1 byte: the pallet index the transaction is calling into.
+  - 1 byte: the function in the pallet the transaction is calling.
+  - variable: the SCALE-encoded parameters required by the function being called.
```
```suggestion
  - variable: the SCALE-encoded parameters required by the function being called.
- 4 bytes: Spec version of the runtime.
- 4 bytes: Transaction version of the runtime.
- 32 bytes: Genesis hash of the chain.
- 32 bytes: Block hash serving as the era reference. If the transaction is immortal, then this would be the genesis hash.
```

joepetrowski commented on docs/build/build-transaction-construction.md:
```diff
@@ -35,6 +35,29 @@ function from the Balances pallet will take:
+- If there is a signature:
+  - a SCALE encoded `sp_runtime::MultiAddress::Id<AccountId32, u32>` indicating the signer(s) of the transaction.
+  - a SCALE encoded `sp_runtime::MultiSignature::{SigningScheme}` with the signature\*.
+  - a SCALE encoded `sp_runtime::generic::Era` indicating for how long this transaction is valid.
```
Can @Imod7 or @tarikgul comment on encoding for mortal vs. immortal transactions?

TarikGul replied:
For encoding between an `Immortal` and `Mortal` era, the following `impl` is useful: https://github.com/paritytech/substrate/blob/master/primitives/runtime/src/generic/era.rs#L102

An immortal era will default to the encoded value of `0x00`. It will be just a single byte of `0`.

A mortal era on the other hand will be `Vec[u64, u64]`; Comprising of the Period, and Phase.

joepetrowski reviewed (commented):
Left some additions, then I think it's just missing signed extensions? Once those are addressed then I think it's ready.

IkerAlus committed:
Joe's suggestion

Co-authored-by: joe petrowski <25483142+joepetrowski@users.noreply.github.com>

IkerAlus committed:
Add signed extensions point and `sp_runtime::generic::Era` encoding info.

It would be nice if @joepetrowski double checks the point about the signed extensions. It is not entirely clear to me how this info should be referenced here for serialization.

IkerAlus commented:
> Left some additions, then I think it's just missing signed extensions? Once those are addressed then I think it's ready.

I added the info for the era encoding and the signed extensions point in my last commit. It would be nice if @joepetrowski double checks the point about the signed extensions. It is not entirely clear to me how this info should be referenced here for serialization.

joepetrowski reviewed (approved):
Would just run `prettier` for formatting, then it's good!

IkerAlus committed:
Run Prettier
